tests/testthat/test_roles.R

```diff
@@ -27,31 +27,284 @@ test_that('changing roles', {
-    rec <- add_role(rec, sample, new_role = "other other role")
+    rec <- add_role(rec, sample, new_role = NA)
   )
+
   exp_res <- tibble(variable = colnames(biomass),
                     type = rep(c("nominal", "numeric"), c(2, 6)),
-                    role = rep(c("other other role", NA), c(1, 7)),
+                    role = NA,
+                    source = "original")
+  expect_equal(summary(rec, TRUE), exp_res)
+
+})
+
+test_that('NA role can be added', {
+
+  rec <- recipe(x = biomass)
+  rec <- update_role(rec, sample, new_role = "role1")
+  rec <- add_role(rec, sample, new_role = NA)
+
+  exp_res <- tibble(variable = c("sample", colnames(biomass)),
+                    type = rep(c("nominal", "numeric"), c(3, 6)),
+                    role = c("role1", rep(NA, 8)),
                     source = "original")
   expect_equal(summary(rec, TRUE), exp_res)
+
 })
 
+test_that('New type for an existing role can be added', {
+
+  rec <- recipe(x = biomass)
+  rec <- update_role(rec, sample, new_role = "role1")
+  rec <- add_role(rec, sample, new_role = "role1", new_type = "text")
+
+  exp_res <- tibble(variable = c("sample", colnames(biomass)),
+                    type = c(c("nominal", "text", "nominal"), rep("numeric", 6)),
+                    role = c("role1", "role1", rep(NA, 7)),
+                    source = "original")
+  expect_equal(summary(rec, TRUE), exp_res)
+
+})
 
 test_that('bad args', {
   expect_error(
-    expect_warning(
-      recipe(x = biomass) %>%
-        add_role(carbon, new_role = letters[1:2])
-    )
+    recipe(x = biomass) %>%
+      add_role(carbon, new_role = letters[1:2]),
+    "`new_role` must have length 1."
   )
+
   expect_error(
     expect_warning(
       recipe(x = biomass) %>%
         add_role(new_role = "some other role")
-    )
+    ),
+    "At least one selector"
   )
+
+  expect_error(
+    recipe(x = biomass) %>%
+      add_role(carbon, new_role = "a", new_type = letters[1:2]),
+    "`new_type` must have length 1."
+  )
+
+  expect_error(
+    recipe(x = biomass) %>%
+      update_role(carbon, new_role = c("a", "b")),
+    "`new_role` must have length 1."
+  )
+
+  expect_error(
+    recipe(x = biomass) %>%
+      update_role(carbon, old_role = c("a", "b")),
+    "`old_role` must have length 1."
+  )
+
+})
+
+
+# ------------------------------------------------------------------------------
+# Multiples roles + Selection testing
+
+test_that("adding multiple roles/types does not duplicate prepped columns", {
+
+  rec <- recipe(HHV ~ ., data = biomass)
+
+  # second role
+  expect_equal(
+    rec %>%
+      add_role(carbon, new_role = "carb") %>%
+      prep(training = biomass, retain = TRUE) %>%
+      juice() %>%
+      ncol(),
+
+    8
+  )
+
+  # second type
+  expect_equal(
+    rec %>%
+      add_role(carbon, new_type = "carb") %>%
+      prep(training = biomass, retain = TRUE) %>%
+      juice() %>%
+      ncol(),
+
+    8
+  )
+
+})
+
+test_that("type selectors can be combined", {
```

Comment:
I'm getting a warning during `test()` that there is no `expect_*` here.

Reply:
Oh whoops ill fix that

Reply:
I can add it to my branch